enhancements/microshift/microshift-updateability-ostree.md

```diff
@@ -0,0 +1,774 @@
+
+%% Decisions
+
+  data-dir-exists?{/var/lib/microshift/}
+  backup-or-restore?{Backup or restore\nbased on previous boot}
+  backup-exists?{Does backup exist?}
+  version?{Compare\nMicroShift's version:\nmetadata vs binary}
+  data-migration?{Was\nmigration\nsuccessful?}
+  healthy?{"Is MicroShift healthy?\n(greenboot check)"}
+  missing-metadata-version?{Upgrade\nfrom 4.13 is\nsupported}
+  restore-ok?{Restore\nsucceeded?}
+  backup-ok?{Backup\nsucceeded?}
+
+%% Transitions
+
+  start --> pre-run --> data-dir-exists?
+
+  data-dir-exists? -- Doesn't exists --> run
+  data-dir-exists? -- Exists         --> backup-or-restore?
+
+  backup-or-restore? -- "Restore (red)"    --> do-restore
+  backup-or-restore? -- "Backup (green)"   --> do-backup
+  backup-or-restore? -- Missing\nData exists,\nbut no info persisted about what to do\nFits upgrade from 4.13 flow --> do-backup
+
+  do-backup --> backup-ok?
+  backup-ok? -- Yes --> version?
+  backup-ok? -. No .-> exit
+
+  do-restore                             --> backup-exists?
+  backup-exists? -- No --> do-clean-data --> run
+  backup-exists? -- Yes --> restore-ok? 
+  restore-ok? -- Yes --> version?
+  restore-ok? -. No .-> exit
+
+  version? -- Versions are the same                 --> run
+  version? -. Binary is older                       .-> exit
+  version? -- Binary is newer                       --> do-data-migration
+  version? -. Upgrade is blocked                    .-> exit
+  version? -- Upgrade is allowed                    --> do-data-migration
+  version? -- "Metadata is missing\n(assume 4.13)"  --> missing-metadata-version?
+
+  missing-metadata-version? -- Yes --> do-data-migration
+  missing-metadata-version? -. No .-> exit
+
+  do-data-migration --> data-migration?
+  data-migration? -- Yes --> run
+  data-migration? -. No  .-> exit
+
+  run --> healthy?
+  healthy? -- Yes --> green -->  continue-running
+  healthy? -- Yes --> do-persist-version
+  healthy? -. No  .-> red
+
+  exit -.-> red
+  red --> reboot
+```
+
+### Test Plan
+
+#### Unit tests
+
+Aiming to write as much as possible in Go, we should strive for maximum testability: 
+- Separate code paths for planning (e.g. should it do a backup or restore?) 
+  and acting (actually perform backup) - e.g. interface with two methods Plan(), Act()
+  - This will allow testing decisions and actions separately
+  - This will allow easy implementation of --dry-run describing what would happen
+- Due to many interactions with filesystem, filesystem abstraction should be investigated
+  so unit tests can use in-memory filesystem rather than host's to make testing easier
+  and more robust.
+
+#### Integration tests focused on each of the areas (backup, restore, migrate)
+
+Ideas:
+Failed backup: `/var/lib/microshift.bak` is not-writable
+Failed restore: `/var/lib/microshift.bak` is not-readable
+
+#### End to end tests
+Following tests should be implemented in CI:
+- TODO
```

Comment:
Most, if not all, of the workflows described above could translate directly to test cases.